api/ai/new/rarest_first_algorithm/rarest_first_algorithm.py

```diff
@@ -0,0 +1,169 @@
+
+        return max_distance_of_each_student
+
+    def _build_shortest_distance_to_attribute_look_up(
+        self,
+        students: List[Student],
+        support_groups: Dict[int, SupportGroup],
+        social_graph: RawSocialGraph,
+    ):
+        distances: Dict[int, Dict[int, Distance]] = {
+            student.id: {} for student in students
+        }
+        for student in students:
+            for attribute_id in self.attributes:
+                current_support_group = support_groups.get(attribute_id)
+                if not current_support_group:
+                    continue
+
+                min_distance = Distance(float("inf"))
+                for other in current_support_group.students:
+                    if other.id == student.id:
+                        continue
+
+                    current_distance = social_graph.get_shortest_distance(
+                        student, other
+                    )
+                    if current_distance < min_distance.value:
+                        min_distance.value = current_distance
+                        min_distance.start_student = student
+                        min_distance.end_student = other
+                        min_distance.is_updated = True
+
+                if min_distance.is_updated:
+                    distances[student.id][attribute_id] = min_distance
+
+        return distances
+
+    def generate(self, students: List[Student]) -> TeamSet:
+        # Set up
+        student_id_trace = {student.id: student for student in students}
+        social_graph = RawSocialGraph(students)
+        support_groups = self._get_students_for_all_attributes(students)
+        least_supported_group = self._get_least_supported_group(support_groups)
+        shortest_distance_to_attribute = (
+            self._build_shortest_distance_to_attribute_look_up(
+                students, support_groups, social_graph
+            )
+        )
+        max_distances = self._get_max_distances(
+            least_supported_group, social_graph, support_groups
+        )
+
+        # Run algorithm
+        min_distance = Distance(float("inf"))  # i_star
+        for student_id in max_distances:
+            current_distance = max_distances[student_id]
+            if current_distance.value < min_distance.value:
+                min_distance = current_distance
+
+        result: Set[int] = {min_distance.start_student.id}
+        for attribute_id in self.attributes:
+            distance_from_i_star = shortest_distance_to_attribute[
+                min_distance.start_student.id
+            ].get(attribute_id)
+            if distance_from_i_star and distance_from_i_star.is_updated:
+                result.add(distance_from_i_star.end_student.id)
+
+        team = Team(
+            name="Subgraph",
+            students=list(student_id_trace.get(student_id) for student_id in result),
+            _id=0,
+        )
+        return TeamSet(teams=[team])
```

Comment:
why only 1 team?

Reply:
This algorithm only outputs 1 team (sub graph in a social network)